Arith & Logic insn

GlacierW opened the pull request:
1. Add support to generate julia code for opcode group
2. Add template for opcode group 0x00~0x35 (the original template of 0x31 is removed)
3. Add template for opcode group 0x08~0x3d
4. Add RFLAG initialization and lazy computation
5. Re-format the CPU dump

GlacierW committed:
1. add 0x00~0x35 ALU instruction template
2. modify gen_instr.jl to support opcode group generation
3. add implicit parameter opc to emu_OPC/jit_OPC function for the opcode parsing

GlacierW committed:
Merge branch 'master' into arith_logic_insn

GlacierW committed:
Merge branch 'master' into arith_logic_insn

GlacierW committed:
add CPU rflag status lazy computation

GlacierW committed:
1. add rflags lazy computation for CMP instruction
2. fix parameter bug of emu/jit_OPCGROUP functions
3. implement another set of arith & logic instruction, opcode 0x08~0x3d

GlacierW committed:
Merge branch 'master' into arith_logic_insn

Conflicts:
	CPU.jl

GlacierW committed:
1. refactor templates of the opcode groups: 0x00-0x35 and 0x08-0x3d
2. fix the bug when computing carry flag of rflags
3. add CPU_XX constants as the mask of the flags of rflags

GlacierW committed:
1. add rflags initilization when resetting CPU
2. re-format the output of CPU dump

GlacierW committed:
1. Omit the use of eval() in opcode templates
2. Fix the bug of CMP instruction, where the actual substraction should not be performed

hackjackhack commented:
Suggestion:
1. Define CPU flag constants with shift operator. (E.g. 0x0000000000000010 -> 0x1 << 4)
2. Use @code_native to check the code of rflags_compute().
3. Remove return in rflags_compute.
4. Add comment to explain the template rule of group expanding.
5. Comment 0x00_0x35.t
6. Fix: rflags_compute! should call rflags_compute_xxx!().  '!' is missed in current code.
7. Fix the bug that ADC with CF = 0 stills sets OP_ADC.

GlacierW committed:
Make the operation type of ADC/SBB instruction be decided in runtime phase for RFLAGS lazy computation

GlacierW committed:
1. Replace the RFLAGS mask with bit-shift coding style
2. Add the rule explanation for writing grouped-opcode templates
3. Rewrite the rflags lazy-computing functions with registered-callback handlers
   Based on the @code_native evaluation, the generated code is much lesser than the previous if-elseif implementation
4. Fix bug of CPU status dump for RFLAGS
5. Fix bug of setting wrong RFLAGS lazy-computing operand of Arith & Logic insn. emulation

For debug use, currently RFALGS will be updated after each instruction execution (in function loop of CPU.jl)

GlacierW commented:
The comments issued previously have been addressed.
Please refer to the git commit message.
